noaa/gefs/forecast/read_data.py

```diff
@@ -27,66 +53,195 @@ def download_file(
-    ds = ds.drop_vars([c for c in ds.coords if c not in ds.dims])
+        byte_range_starts.append(start_byte)
+        byte_range_ends.append(end_byte)
+    return byte_range_starts, byte_range_ends
+
+
+def format_noaa_idx_var(var_info: DataVar) -> str:
+    return f"{var_info.internal_attrs.grib_element}:{var_info.internal_attrs.grib_index_level}"
+
+
+def digest(data: str | Iterable[str], length: int = 8) -> str:
+    """Consistent, likely collision free string digest of one or more strings."""
+    if isinstance(data, str):
+        data = [data]
+    message = hashlib.sha256()
+    for string in data:
+        message.update(string.encode())
+    return message.hexdigest()[:length]
 
-    # Convert longitude from [0, 360) to [-180, +180)
-    ds = ds.assign_coords(longitude=ds["longitude"] - 180)
 
-    return ds
+# def read_file(file_name: str) -> xr.Dataset:
+#     """
+#     Requirement: the process current working directory must contain file_name
+#     """
+#     # Open the grib as N different datasets which correctly map grib to xarray datasets
+#     datasets = cfgrib.open_datasets(file_name)
+#     datasets = [ds.chunk(-1) for ds in datasets]
+
+#     # TODO compat="minimal" is dropping 3 variables
+#     ds = xr.merge(
+#         datasets, compat="minimal", join="exact", combine_attrs="no_conflicts"
+#     )
+
+#     renames = {"time": "init_time", "step": "lead_time"}
+#     ds = ds.expand_dims(tuple(renames.keys())).rename(renames)
+#     if "number" in ds.coords:
+#         ds = ds.expand_dims("number", axis=1).rename(number="ensemble_member")
+#     else:
+#         # Store summary statistics across ensemble members as separate variables
+#         # which do not have an ensemble_member dimension. Statistic is either
+#         # "avg" (ensemble mean) or "spr" (ensemble spread; max minus min)
+#         statistic = file_name[: file_name.index(".")].removeprefix("ge")
+#         ds = ds.rename({var: f"{var}_{statistic}" for var in ds.data_vars.keys()})
+#         for data_var in ds.data_vars.values():
+#             data_var.attrs["long_name"] = (
+#                 f"{data_var.attrs["long_name"]} ({_STATISTIC_LONG_NAME[statistic]})"
+#             )
+
+#     # Drop level height coords which belong on the variable and are not dataset wide
+#     ds = ds.drop_vars([c for c in ds.coords if c not in ds.dims])
+
+#     # Convert longitude from [0, 360) to [-180, +180)
+#     ds = ds.assign_coords(longitude=ds["longitude"] - 180)
+
+#     return ds
+
+
+def read_into(
+    out: xr.DataArray,
+    coords: SourceFileCoords,
+    path: os.PathLike[str],
+    data_var: DataVar,
+) -> None:
+    grib_element = data_var.internal_attrs.grib_element
+    if data_var.internal_attrs.include_lead_time_suffix:
+        lead_hours = coords["lead_time"].total_seconds() / (60 * 60)
+        if lead_hours % 6 == 0:
+            grib_element += "06"
+        else:
+            grib_element += "03"
```

Comment:
lets put this within some sort of afirmative elif check for 03 and have the else raise an AssertionError or similar so we fail in any future cases we aren't thinking about